src/compositor/input/pointer.rs

```diff
@@ -149,8 +158,20 @@ fn focus_under_pointer<'view, V>(seat: &mut compositor::Seat,
             })
         }
         Some(view) => {
+            if let Some(mut focused_view) = seat.focused.as_mut() {
+                if focused_view.shell == view.shell {
+                    return Ok(())
+                }
+                focused_view.activate(false);
+            }
             seat.focused = Some(view.clone());
             view.activate(true);
+
+            if let Some(idx) = views.iter().position(|ref v| v.shell == view.shell) {
+                let v = views.remove(idx);
+                views.insert(0, v);
```

Comment:
So wait, the focused view is always the top of the list? So it's a stack?

I don't really like this, as it's not very performant and seems rather hacky. I also see why you can't just have use a `focused`, since that might differ from what's in the view list.

If upgrading a handle didn't require an `&mut` you could just use `Rc` here and call it a day. I'll have to re-investigate why I made it require `&mut`. I believe I found a safety hole when it was simply `&`, but it might be me being over zealous.

Reply:
[Found the issue where I made it &mut](https://github.com/swaywm/wlroots-rs/pull/76).

It's not UB, I'm not sure what I was talking about. If we switch the handles to only require a `&` to upgrade the handle then, at the cost of making it slightly easier to shoot yourself in the foot (e.g by immediantly re-upgrading from the same handle, which will just cause an error and never be an issue) we can have `Rc<View>` stored at `compositor::Seat.focused`  that can also be inside the `Server.views` list. That way they'll always be synced up. 

For the values inside `View` that still need to be mutated and shared (e.g right now just the `Origin`) we can add `Cell`s (in the case of the `Origin`, which is more ergonomic) or `RefCell`s in the future.

So the view will look like this:


```rust
pub struct View {
    pub shell: Shell,
    /// Allows us to mutate it behind an Rc
    pub origin: Cell<Origin>
}
```

and the seat will look like this:

```rust
pub struct Seat {
    pub seat: SeatHandle,
    /// Thanks to zero cost abstractions, 
    /// this Option<Rc> has no overhead compared to a raw pointer in terms of size.
    pub focused: Option<Rc<View>>,
    pub action: Option<Action>,
    pub meta: bool
}
```

and finally the `Server` will look like this:

```rust
pub struct Server {
    // ... pointless fields elided

    pub views: Vec<Rc<View>>
}
```